File: src/main/java/back/ecommerce/auth/interceptor/JwtAuthenticationInterceptor.java
package back.ecommerce.auth.interceptor;

import static back.ecommerce.exception.ErrorCode.*;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.servlet.HandlerInterceptor;

import back.ecommerce.auth.token.TokenProvider;
import back.ecommerce.exception.AuthenticationException;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class JwtAuthenticationInterceptor implements HandlerInterceptor {

	private static final String AUTHORIZATION_TYPE = "Bearer ";
	private static final String EMAIL_ATTRIBUTE = "email";

	private final TokenProvider tokenProvider;

	@Override
	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws
		Exception {
		String token = extractHeaderToToken(request.getHeader(HttpHeaders.AUTHORIZATION));
		String email = tokenProvider.extractClaim(token, EMAIL_ATTRIBUTE);
		request.setAttribute(EMAIL_ATTRIBUTE, email);
		return true;
	}

	private String extractHeaderToToken(String header) {
		if (!StringUtils.hasText(header)) {
			throw new AuthenticationException(AUTH_HEADER_IS_EMPTY);
		}
		if (!header.startsWith(AUTHORIZATION_TYPE)) {
			throw new AuthenticationException(AUTH_HEADER_INVALID);
		}
		return header.substring(AUTHORIZATION_TYPE.length());
	}
}

File: src/main/java/back/ecommerce/api/auth/interceptor/JwtAuthenticationInterceptor.kt
package back.ecommerce.api.auth.interceptor

import back.ecommerce.auth.service.TokenExtractor
import back.ecommerce.exception.AuthenticationException
import back.ecommerce.exception.ErrorCode
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import lombok.RequiredArgsConstructor
import org.springframework.http.HttpHeaders
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor

@Component
@RequiredArgsConstructor
class JwtAuthenticationInterceptor(
    private val tokenExtractor: TokenExtractor
) : HandlerInterceptor {

    @Throws(Exception::class)
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {

        val header = request.getHeader(HttpHeaders.AUTHORIZATION) ?: throw AuthenticationException(ErrorCode.TOKEN_IS_EMPTY)
        if (!header.startsWith(AUTHORIZATION_TYPE)) {
            throw AuthenticationException(ErrorCode.AUTH_HEADER_INVALID)
        }
        header.removePrefix(AUTHORIZATION_TYPE)
        tokenExtractor.extractClaim(header, EMAIL_ATTRIBUTE).also {
            request.setAttribute(EMAIL_ATTRIBUTE, it)
        }
        return true
    }

    companion object {
        private const val AUTHORIZATION_TYPE = "Bearer "
        private const val EMAIL_ATTRIBUTE = "email"
    }
}
